src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,584 @@
+     * @throws NullPointerException
+     *     if {@code alg} or {@code kdfParameterSpec} is null
+     */
+    public SecretKey deriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        synchronized (lock) {
+            if (alg == null || alg.isEmpty()) {
+                throw new NullPointerException(
+                    "the algorithm for the SecretKey return value may not be "
+                    + "null or empty");
+            }
+            Objects.requireNonNull(kdfParameterSpec);
+            if (spi != null) {
+                return spi.engineDeriveKey(alg, kdfParameterSpec);
+            }
+            Exception lastException = null;
+            while ((firstService != null) || serviceIterator.hasNext()) {
+                Service s;
+                if (firstService != null) {
+                    s = firstService;
+                    firstService = null;
+                } else {
+                    s = serviceIterator.next();
+                }
+                if (!JceSecurity.canUseProvider(s.getProvider())) {
+                    continue;
+                }
+                try {
+                    KDFSpi spi = (KDFSpi) s.newInstance(algorithmParameterSpec);
+                    SecretKey result = spi.engineDeriveKey(alg,
+                                                           kdfParameterSpec);
+                    provider = s.getProvider();
+                    this.spi = spi;
+                    firstService = null;
+                    serviceIterator = null;
+                    return result;
+                } catch (Exception e) {
+                    if (lastException == null) {
+                        lastException = e;
+                    }
+                }
+            }
+            // no working provider found, fail
+            if (lastException instanceof InvalidParameterSpecException) {
+                throw (InvalidParameterSpecException) lastException;
+            }
+            if (lastException instanceof RuntimeException) {
+                throw (RuntimeException) lastException;
+            }
+        }
+        // should never reach here
+        return null;
+    }
+
+    /**
+     * Obtains raw data from a key derivation function.
+     * <p>
+     * The {@code deriveData} method may be called multiple times on a
+     * particular {@code KDF} instance.
+     * <p>
+     * Delayed provider selection is also supported such that the provider
+     * performing the derive is not selected until the method is called. Once a
+     * provider is selected, it cannot be changed.
+     *
+     * @param kdfParameterSpec
+     *     derivation parameters (may not be null)
+     *
+     * @return a byte array whose length matches the specified length in the
+     *     processed {@code KDFParameterSpec} and containing the output from the
+     *     key derivation function
```

Comment:
There is no length parameter in `KDFParameterSpec`. Suggest making this more generic as you did for `deriveKey`.